src/anomalib/metrics/per_image/_validate.py

```diff
@@ -0,0 +1,422 @@
+        one_ok (bool): Flag indicating if rate can be 1.
+    """
+    if not isinstance(rate, float | int):
+        msg = f"Expected rate to be a float or int, but got {type(rate)}."
+        raise TypeError(msg)
+
+    if rate < 0.0 or rate > 1.0:
+        msg = f"Expected rate to be in [0, 1], but got {rate}."
+        raise ValueError(msg)
+
+    if not zero_ok and rate == 0.0:
+        msg = "Rate cannot be 0."
+        raise ValueError(msg)
+
+    if not one_ok and rate == 1.0:
+        msg = "Rate cannot be 1."
+        raise ValueError(msg)
+
+
+def rate_range(bounds: tuple[float, float]) -> None:
+    """Validates the range of rates within the bounds.
+
+    Args:
+        bounds (tuple[float, float]): The lower and upper bounds of the rates.
+    """
+    if not isinstance(bounds, tuple):
+        msg = f"Expected the bounds to be a tuple, but got {type(bounds)}"
+        raise TypeError(msg)
+
+    if len(bounds) != 2:
+        msg = f"Expected the bounds to be a tuple of length 2, but got {len(bounds)}"
+        raise ValueError(msg)
+
+    lower, upper = bounds
+    rate(lower, zero_ok=False, one_ok=False)
+    rate(upper, zero_ok=False, one_ok=True)
+
+    if lower >= upper:
+        msg = f"Expected the upper bound to be larger than the lower bound, but got {upper=} <= {lower=}"
+        raise ValueError(msg)
+
+
+def file_path(file_path: str | Path, must_exist: bool, extension: str | None, pathlib_ok: bool) -> None:
+    """Validate the given path is a file (optionally) with the expected extension.
+
+    Args:
+        file_path (str | Path): The file path to validate.
+        must_exist (bool): Flag indicating whether the file must exist.
+        extension (str | None): The expected file extension, eg. .png, .jpg, etc. If `None`, no validation is performed.
+        pathlib_ok (bool): Flag indicating whether `pathlib.Path` is allowed; if False, only `str` paths are allowed.
+    """
+    if isinstance(file_path, str):
+        file_path = Path(file_path)
+
+    elif not isinstance(file_path, Path):
+        msg = f"Expected file path to be a string or pathlib.Path, but got {type(file_path)}"
+        raise TypeError(msg)
+
+    # if it's here, then it's a `pathlib.Path`
+    elif not pathlib_ok:
+        msg = f"Only `str` paths are allowed, but got {type(file_path)}"
+        raise TypeError(msg)
+
+    if file_path.is_dir():
+        msg = "Expected file path to be a file, but got a directory."
+        raise ValueError(msg)
+
+    if must_exist and not file_path.exists():
+        msg = f"File does not exist: {file_path}"
+        raise FileNotFoundError(msg)
+
+    if extension is None:
+        return
+
+    if file_path.suffix != extension:
+        msg = f"Expected file path to have extension '{extension}', but got '{file_path.suffix}'"
+        raise ValueError(msg)
+
+
+def file_paths(file_paths: list[str | Path], must_exist: bool, extension: str | None, pathlib_ok: bool) -> None:
```

Comment:
so the idea of  the `_validate` module is to have short names because the "validate" part is common to all the methods; the functions are always called with `_validate.file_paths()` to make the "validate" show up

sounds logical?